torch/export/_trace.py

```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
+        return kernel(*args, **kwargs)
+
+
+def _register_cia_to_meta(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(
+            torch._C.DispatchKey.CompositeImplicitAutograd, *args, **kwargs
+        )
+    else:
+        return NotImplemented
+
+
+@contextmanager
+def override_composite_implicit_decomp(ops_to_preserve):
+    # This function overrides CompositeImplicitAutograd decomp for
+    # functional composite ops. Ideally we want to not-decompose
+    # ALL composite ops but today's C++ functinalization relies on
+    # the fact that it is working with the opset after decomp is run.
+    # Hence we can only do it for functional ops. One caveat is that
+    # there are some composite ops that lie about their schema (claimed to be
+    # functional but not really aka dropout), for these cases, we just decompose.
+    saved_tables = {}
+    patched_ops = set()
+    for op_overload in ops_to_preserve:
+
+        def can_preserve(op_overload):
+            if op_overload in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+                return False
+            if op_overload in FunctionalTensor.metadata_fns:
+                return False
+
+            alias_info = len(
+                [i for i in op_overload._schema.arguments if i.alias_info is not None]
+            )
+
+            is_mutating_or_aliasing = alias_info != 0 or op_overload._schema.is_mutable
+
+            if is_mutating_or_aliasing:
+                return False
+
+            if not torch._C._dispatch_has_kernel(op_overload.name()):
+                return False
+
+            if not torch._C._dispatch_has_kernel_for_dispatch_key(
+                op_overload.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+            ):
+                return False
+
+            return True
+
+        if not can_preserve(op_overload):
+            warnings.warn(
+                f"We can't preserve {op_overload} in export because this op is unsafe to keep"
+            )
+            continue
+
+        saved_tables[op_overload] = op_overload.py_kernels.copy()
+        patched_ops.add(op_overload)
+        for override_dispatch_key in [
+            torch._C.DispatchKey.AutogradCPU,
+            torch._C.DispatchKey.AutogradCUDA,
+            torch._C.DispatchKey.AutogradMeta,
+        ]:
+            if override_dispatch_key not in op_overload.py_kernels:
+                # conv1d, conv2d, conv3d don't work with default Autograd key
+                op_overload.py_impl(override_dispatch_key)(
+                    functools.partial(_skip_autograd_kernel, kernel=op_overload)
+                )
+        if torch._C.DispatchKey.CompositeImplicitAutograd in op_overload.py_kernels:
+            del op_overload.py_kernels[torch._C.DispatchKey.CompositeImplicitAutograd]
```

Comment:
this doesn't feel very safe.

In your finally block, you probably want to undo all of the dispatch table changes you made so far (swapping out the Meta and CompositeImplicitAutograd registrations)

Reply:
Yep i save the table before this and set it afterwards.